File: doodad-runtime/tools/doodad_cli/canvas_display_list.py
from __future__ import annotations

import re

from .contract import DoodadError


MAX_CANVAS_EDGE = 192
MAX_DISPLAY_LIST_LENGTH = 128
MAX_PALETTE_LENGTH = 64
MAX_PALETTE_COLORS = 8
MAX_TILE_COUNT = 64

_PALETTE = re.compile(r"^[0-9a-f]{6}(?:,[0-9a-f]{6}){0,7}$")


def validate_canvas(
    display_list: object,
    palette: object,
    width: object,
    height: object,
    path: str,
) -> None:
    if (
        not isinstance(width, int)
        or isinstance(width, bool)
        or not 1 <= width <= MAX_CANVAS_EDGE
    ):
        raise DoodadError(f"{path}.width must be in 1..{MAX_CANVAS_EDGE}")
    if (
        not isinstance(height, int)
        or isinstance(height, bool)
        or not 1 <= height <= MAX_CANVAS_EDGE
    ):
        raise DoodadError(f"{path}.height must be in 1..{MAX_CANVAS_EDGE}")
    if (
        not isinstance(palette, str)
        or not 6 <= len(palette) <= MAX_PALETTE_LENGTH
        or _PALETTE.fullmatch(palette) is None
    ):
        raise DoodadError(
            f"{path}.palette must contain 1..{MAX_PALETTE_COLORS} "
            "lowercase RGB888 colors"
        )
    if (
        not isinstance(display_list, str)
        or not 1 <= len(display_list) <= MAX_DISPLAY_LIST_LENGTH
    ):
        raise DoodadError(
            f"{path}.display_list must contain "
            f"1..{MAX_DISPLAY_LIST_LENGTH} characters"
        )

    palette_size = palette.count(",") + 1
    commands = display_list.split("|")
    if not commands or commands[0] != "v1" or len(commands) == 1:
        raise DoodadError(f"{path}.display_list must begin with v1")
    if not commands[1].startswith("C"):
        raise DoodadError(f"{path}.display_list must clear before drawing")
    if len(commands) > 33:
        raise DoodadError(f"{path}.display_list has too many commands")

    for index, command in enumerate(commands[1:]):
        command_path = f"{path}.display_list command {index}"
        if not command:
            raise DoodadError(f"{command_path} is empty")
        opcode = command[0]
        arguments = command[1:].split(",")
        if opcode == "C":
            _expect(arguments, 1, command_path)
            _palette_index(arguments[0], palette_size, command_path)
        elif opcode == "R":
            _expect(arguments, 6, command_path)
            _palette_index(arguments[0], palette_size, command_path)
            x, y, rect_width, rect_height, radius = [
                _number(value, command_path) for value in arguments[1:]
            ]
            if (
                rect_width == 0
                or rect_height == 0
                or x + rect_width > width
                or y + rect_height > height
                or radius > min(rect_width, rect_height) // 2
            ):
                raise DoodadError(f"{command_path} is outside the canvas")
        elif opcode == "O":
            _expect(arguments, 4, command_path)
            _palette_index(arguments[0], palette_size, command_path)
            center_x, center_y, radius = [
                _number(value, command_path) for value in arguments[1:]
            ]
            if (
                radius == 0
                or center_x < radius
                or center_y < radius
                or center_x + radius > width
                or center_y + radius > height
            ):
                raise DoodadError(f"{command_path} is outside the canvas")
        elif opcode == "L":
            _expect(arguments, 6, command_path)
            _palette_index(arguments[0], palette_size, command_path)
            x1, y1, x2, y2, stroke = [
                _number(value, command_path) for value in arguments[1:]
            ]
            if (
                x1 >= width
                or x2 >= width
                or y1 >= height
                or y2 >= height
                or not 1 <= stroke <= 16
            ):
                raise DoodadError(f"{command_path} is outside the canvas")
        elif opcode == "T":
            _expect(arguments, 8, command_path)
            inset, x, y, cell_width, cell_height, columns, rows = [
                _number(value, command_path) for value in arguments[:7]
            ]
            cells = arguments[7]
            cell_count = columns * rows
            if (
                cell_width == 0
                or cell_height == 0
                or columns == 0
                or rows == 0
                or cell_count > MAX_TILE_COUNT
                or len(cells) != cell_count
                or inset * 2 >= min(cell_width, cell_height)
                or x + cell_width * columns > width
                or y + cell_height * rows > height
                or any(
                    character < "0"
                    or character > "7"
                    or int(character) >= palette_size
                    for character in cells
                )
            ):
                raise DoodadError(f"{command_path} has an invalid tile map")
        else:
            raise DoodadError(f"{command_path} uses unsupported opcode {opcode!r}")


def _expect(arguments: list[str], count: int, path: str) -> None:
    if len(arguments) != count:
        raise DoodadError(f"{path} has the wrong number of arguments")


def _number(value: str, path: str) -> int:
    if not value or not value.isascii() or not value.isdecimal():
        raise DoodadError(f"{path} contains a non-decimal argument")
    number = int(value)
    if number > MAX_CANVAS_EDGE:
        raise DoodadError(f"{path} contains an out-of-range argument")
    return number


def _palette_index(value: str, palette_size: int, path: str) -> int:
    number = _number(value, path)
    if number >= palette_size:
        raise DoodadError(f"{path} uses a missing palette color")
    return number

File: doodad-runtime/reference/android-wear/app/src/main/java/dev/doodad/reference/model/CanvasDisplayList.kt
package dev.doodad.reference.model

data class ParsedCanvasDisplayList(
    val palette: List<Int>,
    val commands: List<CanvasCommand>,
)

sealed interface CanvasCommand {
    data class Clear(
        val color: Int,
    ) : CanvasCommand

    data class RoundedRect(
        val color: Int,
        val x: Int,
        val y: Int,
        val width: Int,
        val height: Int,
        val radius: Int,
    ) : CanvasCommand

    data class Circle(
        val color: Int,
        val centerX: Int,
        val centerY: Int,
        val radius: Int,
    ) : CanvasCommand

    data class Line(
        val color: Int,
        val x1: Int,
        val y1: Int,
        val x2: Int,
        val y2: Int,
        val stroke: Int,
    ) : CanvasCommand

    data class TileMap(
        val inset: Int,
        val x: Int,
        val y: Int,
        val cellWidth: Int,
        val cellHeight: Int,
        val columns: Int,
        val rows: Int,
        val cells: List<Int>,
    ) : CanvasCommand
}

object CanvasDisplayListCodec {
    fun parse(
        displayList: String,
        palette: String,
        width: Int,
        height: Int,
    ): ParsedCanvasDisplayList {
        check(width in 1..192 && height in 1..192) {
            "Canvas dimensions must be in 1..192"
        }
        check(displayList.length in 1..128) {
            "Canvas Display List must contain 1..128 characters"
        }
        check(
            palette.length in 6..64 &&
                palette.matches(Regex("^[0-9a-f]{6}(,[0-9a-f]{6}){0,7}$")),
        ) {
            "Canvas palette must contain 1..8 lowercase RGB888 colors"
        }
        val colors =
            palette.split(",").map { encoded ->
                encoded.toInt(16)
            }
        val encodedCommands = displayList.split("|")
        check(
            encodedCommands.size in 2..33 &&
                encodedCommands.first() == "v1" &&
                encodedCommands[1].startsWith("C"),
        ) {
            "Canvas Display List must begin with v1 and clear"
        }
        var cleared = false
        val commands =
            encodedCommands.drop(1).mapIndexed { index, encoded ->
                check(encoded.isNotEmpty()) {
                    "Canvas command $index is empty"
                }
                val opcode = encoded.first()
                val arguments = encoded.drop(1).split(",")
                when (opcode) {
                    'C' -> {
                        check(!cleared && index == 0 && arguments.size == 1)
                        cleared = true
                        CanvasCommand.Clear(
                            colorIndex(arguments[0], colors.size),
                        )
                    }
                    'R' -> {
                        check(arguments.size == 6)
                        val values = arguments.map(::number)
                        val color = colorIndex(arguments[0], colors.size)
                        val x = values[1]
                        val y = values[2]
                        val rectWidth = values[3]
                        val rectHeight = values[4]
                        val radius = values[5]
                        check(
                            rectWidth > 0 &&
                                rectHeight > 0 &&
                                x + rectWidth <= width &&
                                y + rectHeight <= height &&
                                radius <= minOf(rectWidth, rectHeight) / 2,
                        )
                        CanvasCommand.RoundedRect(
                            color,
                            x,
                            y,
                            rectWidth,
                            rectHeight,
                            radius,
                        )
                    }
                    'O' -> {
                        check(arguments.size == 4)
                        val values = arguments.map(::number)
                        val color = colorIndex(arguments[0], colors.size)
                        val centerX = values[1]
                        val centerY = values[2]
                        val radius = values[3]
                        check(
                            radius > 0 &&
                                centerX >= radius &&
                                centerY >= radius &&
                                centerX + radius <= width &&
                                centerY + radius <= height,
                        )
                        CanvasCommand.Circle(
                            color,
                            centerX,
                            centerY,
                            radius,
                        )
                    }
                    'L' -> {
                        check(arguments.size == 6)
                        val values = arguments.map(::number)
                        val color = colorIndex(arguments[0], colors.size)
                        check(
                            values[1] < width &&
                                values[3] < width &&
                                values[2] < height &&
                                values[4] < height &&
                                values[5] in 1..16,
                        )
                        CanvasCommand.Line(
                            color,
                            values[1],
                            values[2],
                            values[3],
                            values[4],
                            values[5],
                        )
                    }
                    'T' -> {
                        check(arguments.size == 8)
                        val values = arguments.take(7).map(::number)
                        val inset = values[0]
                        val x = values[1]
                        val y = values[2]
                        val cellWidth = values[3]
                        val cellHeight = values[4]
                        val columns = values[5]
                        val rows = values[6]
                        val cellData = arguments[7]
                        val cellCount = columns * rows
                        check(
                            cellWidth > 0 &&
                                cellHeight > 0 &&
                                columns > 0 &&
                                rows > 0 &&
                                cellCount <= 64 &&
                                cellData.length == cellCount &&
                                inset * 2 < minOf(cellWidth, cellHeight) &&
                                x + cellWidth * columns <= width &&
                                y + cellHeight * rows <= height,
                        )
                        val cells =
                            cellData.map { character ->
                                colorIndex(character.toString(), colors.size)
                            }
                        CanvasCommand.TileMap(
                            inset,
                            x,
                            y,
                            cellWidth,
                            cellHeight,
                            columns,
                            rows,
                            cells,
                        )
                    }
                    else -> error("Unsupported Canvas opcode $opcode")
                }
            }
        return ParsedCanvasDisplayList(colors, commands)
    }

    private fun number(encoded: String): Int {
        check(encoded.isNotEmpty() && encoded.all { it in '0'..'9' })
        return encoded.toInt().also {
            check(it in 0..192)
        }
    }

    private fun colorIndex(encoded: String, count: Int): Int =
        number(encoded).also {
            check(it in 0 until count)
        }
}
